### ISSUE TITLE
Midroll ads not running when using IMA extension

### ISSUE THREAD

**Issue by khat33b** [google/ExoPlayer#6645]
### [REQUIRED] Searched documentation and issues
I have searched on the GitHub home page, the developer website, past issues and StackOverflow.

### [REQUIRED] Question
I am using a custom implementation of ExoPlayer's IMA extension to display ads at custom intervals not specified by the Vast Tag as mentioned [here](https://stackoverflow.com/questions/49871348/how-to-show-ads-at-specific-points-using-exoplayers-ima-extension). The tag I am using is [this](http://search.spotxchange.com/vast/3.0/85394). What I have done is that I have replaced the cue points from `adsManager` to my custom cue points.

```
if (this.adCuePoints != null)
  adPlaybackState = new AdPlaybackState(getAdGroupTimesUs(this.adCuePoints));
else
  adPlaybackState = new AdPlaybackState(getAdGroupTimesUs(adsManager.getAdCuePoints()));
```

The ads are being displayed on pre-roll but when the ad is being called at mid-roll, no ad is being displayed, only a loader. There is also no log in AdEventListener and AdErrorListener. How do I fix this?

**Comment by marcbaechinger** [comment]
It's difficult to understand what happens from the information you are providing.

Where is the code living which you pasted above and what is the difference between what you are doing and the adsManager is doing?

Further, can you please capture a bug report and upload it here? Best is to add an EventLogger to your player so we can see what is going on in the player.

**Comment by khat33b** [comment]
I have implemented my own `ImaAdsLoader`. I copied the original from the library, added a `adCuePoints` and wherever there is `adPlaybackState = new AdPlaybackState(getAdGroupTimesUs(adsManager.getAdCuePoints()));`, 

I have replaced it with the above code. I have uploaded my custom loader, the log from EventLogger and the bug report.
[MyImaAdsLoader.txt](https://github.com/google/ExoPlayer/files/3841304/MyImaAdsLoader.txt)
[bugreport-BRAVIA_ATV2_PA-OPR2.170623.027.S25-2019-11-13-18-44-27.zip](https://github.com/google/ExoPlayer/files/3841306/bugreport-BRAVIA_ATV2_PA-OPR2.170623.027.S25-2019-11-13-18-44-27.zip)
[log.txt](https://github.com/google/ExoPlayer/files/3841316/log.txt)

The ad is coming in pre-roll but not in mid-roll although when I tried out the IMA Sample Tags, it is working properly when using tags like _VMAP - Pre-roll Single Ad, Mid-roll Standard Pod with 3 ads, Post-roll Single Ad_ but not when using tags like _Single Inline Linear_. I believe this to be the problem that my vast tag also has a single ad which I have to show multiple times.

**Comment by marcbaechinger** [comment]
If you are not creating your own AdsLoader but you are using the ImaAdsLoader instead, then you are probably bound to the logic of the ImaAdsLoader. In general, it appears to me that playing non-IMA ad tags with the ImaAdsLoader could be risky, as you don't know what the ImaAdsLoader expects to find in the IMA specific tag.

If you are just setting more cuepoints then the ImaAdsLoader finds ads in the ad tag the ImaAdsLoader probably never considers the number of queue points as it wants to only show the single ad found in the ad tag. This would be consistent with the fact that it works when number of ads and the cue points you are setting are equal.

I think you don't get around either creating your own AdsLoader or alternatively, you need to know pretty well how the ImaAdsLoader works. For this case it appears that you need to duplicate the

Do you really want to play non-IMA ad tags with the IMA extension? As far as I understand you will send information about ads being displyed, skipped and the like to IMA instead of to your own server.

**Comment by khat33b** [comment]
Oh, I didn't know that `ImaAdsLoader` was fine tuned for only IMA tags, I though it used the IMA SDK and could play well with any VAST tags. I used it because I didn't want to manually initialize and request ads and sync between the video player and ad player and thought the ImaAdsLoader would do all that stuff. I will try to implement my own loader.

**Comment by marcbaechinger** [comment]
I see your point regarding reusing IMA for developer economics.

I think it make sense doing your own here, as it is more efficient if you don't need the IMA spcific parts. IMA is using a webview underneath and does vast/vmap parsing in JavaScript. This ads some memory requirements to your app. If you do this in Java/Kotlin you don't need to communicate with IMA under the hood. You also remove the risk that the JavaScript code in the IMA SDK suddenly does something you did not anticipate. The JavaScript can change without you know because it's loaded each time you use it in your app.

Basically, you do your own AdsLoader implementation which parses vmap/vast files to evaluate what ads are in there. The create a AdsPlaybackState which you pass to adMediaSource.onAdPlaybackState each time something changes in the AdsPlaybackState (like mark ads as being played).

**Comment by marcbaechinger** [comment]
I'm closing this issue for now. Feel free to re-open or create a new one if required.

**Comment by andrewlewis** [comment]
I agree with Marc on "If you are not creating your own AdsLoader but you are using the ImaAdsLoader instead, then you are probably bound to the logic of the ImaAdsLoader.". In particular the IMA SDK tracks the playback position and expects VMAP midrolls to be played at the cue points specified in the ad tag, which is probably why your custom implementation isn't working.

However, I don't think there is a concept of an IMA or "non-IMA" ad tag. The IMA SDK should support standard VAST/VMAP ad tags (assuming they are [compatible](https://developers.google.com/interactive-media-ads/docs/sdks/android/compatibility)).

One option that might work for your use case would be to download the ad tag in your app, rewrite the cue points in the XML, then provide the rewritten document as an ads response (see `ImaAdsLoader.Builder.buildForAdsResponse`). I haven't tried this but I think it'll work it.

**Comment by khat33b** [comment]
Thanks for all the explanations. It helped a lot. In the end, I created my own AdsLoader and it is working fine.
